New guide for ingestion sampling (#16638)
DataDog/documentation: content/en/tracing/guide/ingestion_sampling_use_cases.md

This guide helps you understand when and how to use ingestion control configurations depending on the main use cases you might encounter. 

Ingestion mechanisms are controllable from the Datadog Agent and from Datadog tracing libraries. If you are using OpenTelemetry SDKs to instrument your applications, read [Ingestion Sampling with OpenTelemetry][2].

## Knowing which ingestion mechanisms are used

To identify what ingestion mechanisms are currently used in your Datadog environment, navigate to the [Ingestion Control Page][3].

{{< img src="/tracing/guide/ingestion_sampling_use_cases/ingestion_control_page.png" alt="Ingestion Control Page" style="width:90%;" >}}

The table gives insights on ingested volumes *by service*. The Configuration column provides a first indication of the current set up. It shows: 
- `AUTOMATIC` if the sampling rate calculated in the Datadog Agent is applied to the traces that start from the service. Read more about the specifics of [Datadog Agent ingestion logic][5].
- `CONFIGURED` if a custom sampling rate configured in the tracing library is applied to the traces that start from the service.

> Click on various services to see details about what sampling decision makers are used for each service, as well as what [ingestion sampling mechanisms][1] are leveraged for ingested spans' services. 

Reviewer: Here we reference `sampling decision makers`, but there is no glossary/explanation on what a decision maker is. So, someone new to Datadog concepts will not really know what this is. At least it sounded weird to me (and still sometimes is :) ) when I joined.

Reviewer 2: I added a little clarification :)

The table gives insights on ingested volumes *by service*. The Configuration column provides a first indication of the current set up. It shows: 
- `AUTOMATIC` if the sampling rate calculated in the Datadog Agent is applied to the traces that start from the service. Read more about the specifics of [Datadog Agent ingestion logic][5].

> - `CONFIGURED` if a custom sampling rate configured in the tracing library is applied to the traces that start from the service.

Reviewer 3: Wondering, does it show up as `CONFIGURED` if span sampling is used? Or only trace sampling? Should we say here?

Author: it's for traces only. Good call out

## Configuring ingestion sampling to keep visibility over relevant traces and spans

### Keeping entire transaction traces

> #### Why ? 

Author: Oops, that's french syntax 😬

> #### Why ? 

Reviewer 4: +1, it looks like it's still here and throughout

> #### Why ? 

Reviewer 2: I took out all the question marks 📝

{{< img src="/tracing/guide/ingestion_sampling_use_cases/head_based_sampling_keep.png" alt="Head-based Sampling" style="width:100%;" >}}

To decide which traces to keep and drop, the Datadog Agent computes default sampling rates for each service to apply at trace creation, based on the application traffic:
- For low-traffic applications, a sampling rate of 100% is applied.
- For high-traffic applications, a lower sampling rate is applied with a target of 10 complete traces per second per Agent.

You can also override the default Agent sampling rate by configuring the sampling rate by service. See how to [keep more traces for specific services](#keeping-more-traces-for-high-visibility-into-specific-services-or-resources) for more information.

#### Configuring head-based sampling

Default sampling rates are calculated to target 10 complete traces per second per Agent. You can increase or decrease this target by configuring the Datadog Agent parameter `max_traces_per_second` or the environment variable `DD_APM_MAX_TPS`. Read more about [head-based sampling ingestion mechanisms][5].

> **Note:** Changing an Agent configuration impacts the percentage sampling rates for *all services* reporting traces to this Datadog Agent.

Reviewer 3: It's still per a service though right? Like `service A` send 10 traces that second, `service B` send 7, all of them are still taken right? Maybe this should be clarified here?

Reviewer 2: so line 57 should say "10 complete traces per service per second per Agent," yeah?

Reviewer 3: Yeah

**Note:** Distributed pieces of the trace chunks might not be ingested as the sampling happens locally at the Datadog Agent level.

#### How to configure error sampling ?

You can configure the number of error chunks per second per Agent that you want to capture by setting the environment variable `DD_APM_ERROR_TPS`. The default value is set to `10` errors per second.

To ensure to ingest **all errors**, set the Agent parameter to an arbitrary high value. To disable the error sampler, set `DD_APM_ERROR_TPS` to `0`.

## Reducing ingestion for high volume services

### Reducing volume fromm database or cache services

#### Why ? 

Traced database calls may represent a large amount of ingested data while the application performance metrics (errors, request hits, latency…) are enough to monitor database health.

#### What solutions does Datadog provide ?

In order to reduce the span volume created by tracing database calls, sampling must be configured at the head of the trace. 

Database services are very rarely starting a trace. Most of the time, client database spans are children of an instrumented backend service span. 

> To know **which services start traces** where a database is called, leverage the `Top Sampling Decision Makers` top list graph of the ingestion control page [Service Ingestion Summary][7]. Configuring head-based sampling for these specific services will reduce the volume of ingested database spans, while making sure that no incomplete traces are ingested: distributed traces are either kept or dropped altogether.

Reviewer 3: wow `Top Sampling Decision Makers` is a great feature I didn't know about!

#### How to configure sampling to drop database spans ?

Refer to the [sampling rule configuration section](#how-to-configure-a-sampling-rule) for more information about sampling rules syntax.

For example, if a backend service `my-service` is calling a postgresql database multiple times per trace, which is creating a lot of unwanted span volume: 

- configure a trace sampling rule for the backend service `my-service`. This will ensure 10% of entire traces are kept, including postgresql spans.

```
DD_TRACE_SAMPLING_RULES=[{"service": "my-service", "sample_rate": 0.1}]
```

> - _\[Optional\]_ configure a **single span sampling rule** to keep 100% of the spans for the backend service `my-service`, for instance if some [span-based metrics][8] are built on top of `my-service` data.

Reviewer 3: I think we may also want to make a note, either here, or in the span-based metrics doc, that this will not affect the metrics we generate from spans, e.g. hits, errors, latency, etc. I think users may get confused by this nuance.

In order to optimize visibility into trace data and cut down the noise, Datadog [Ingestion Control Mechanisms][1] helps you keep the visibility that you need to troubleshoot problems while remaining within budget.

This guide helps you understand when and how to use ingestion control configurations depending on the main use cases you might be encountering. 

These Ingestion Mechanisms are controllable from the Datadog Agent and from Datadog tracing libraries. If you are using OpenTelemetry SDKs to instrument your applications, you can refer to the dedicated guide on [Ingestion Sampling with OpenTelemetry][2].

## Knowing which ingestion mechanisms are used

In order to identify what current ingestion mechanisms are used in your Datadog environment, navigate to  the [Ingestion Control Page][3].

{{< img src="/tracing/guide/ingestion_sampling_use_cases/ingestion_control_page.png" alt="Ingestion Control Page" style="width:90%;" >}}

The table gives insights on ingested volumes **by service:**

The configuration column indicating `CONFIGURED` or `AUTOMATIC` provides a first indication of the current set up : 

> - `AUTOMATIC` if the sampling rate calculated in the Datadog Agent is applied for the traces starting from the service. Read more about the specifics of the Datadog Agent logic in the documentation.

Reviewer 5: `Datadog Agent logic in the documentation.`
Is there a missing link?

- `AUTOMATIC` if the sampling rate calculated in the Datadog Agent is applied to the traces that start from the service. Read more about the specifics of [Datadog Agent ingestion logic][5].
- `CONFIGURED` if a custom sampling rate configured in the tracing library is applied to the traces that start from the service.

Click on services to see details about what sampling decision makers (for example Agent or tracing library, rules or sample rates) are used for each service, as well as what [ingestion sampling mechanisms][1] are leveraged for ingested spans' services. 

{{< img src="/tracing/guide/ingestion_sampling_use_cases/service-ingestion-summary.png" alt="Service Ingestion Summary" style="width:90%;" >}}

> 

Reviewer 6: What do you think about explaining this picture out more to really break out how to understand this view?

---

**Explanation of the Example**: 

In the Service Ingestion Summary example above, the **Ingestion reasons breakdown** table shows that most of the ingestion reasons for this service come from **rule** (user defined rule). 

The **Top sampling decision makers for this service** show that **web-store** service gets sampling decisions from **web-store**, **shopist-web-ui**, **shipping-worker**, **synthetics-browser**, and **product-recommendation**.

These five services all contribute in the overall sampling decisions that affect the **web-store** service. When determining how to fine tune the ingestion for **web-store**, all five services need to be considered.

Author: Good idea, I added that here !

### Keeping entire transaction traces

#### Why ? 

Ingesting entire transaction traces ensures to keep visibility over the **end-to-end service request flow** for specific individual requests.

#### What solutions does Datadog provide ?

> Complete traces can be ingested with [head-based sampling][4] mechanisms: the decision of keeping or dropping the trace is taken at the trace creation (trace head), and that this decision is propagated in the request context to downstream callee services.

Reviewer 5: ```suggestion
Complete traces can be ingested with [head-based sampling][4] mechanisms: the decision of keeping or dropping the trace is taken at the trace creation (trace root span), and this decision is propagated in the request context to downstream callee services.
```

{{< img src="/tracing/guide/ingestion_sampling_use_cases/head_based_sampling_keep.png" alt="Head-based Sampling" style="width:100%;" >}}

> To decide which requests to keep and drop, the Datadog Agent computes automatic sampling rates for each service to apply at the root of the trace.

Reviewer 5: This sentence is hard to understand. Do you want to say that the automatic sampling rate set by the agent is the default behaviour but can be over-ridden if necessary?

Author: changed the wording

> To decide which traces to keep and drop, the Datadog Agent computes default sampling rates for each service to apply at trace creation, based on the application traffic:

Reviewer 6: It would be useful to add a link back to this section about **DD_APM_MAX_TPS**: https://docs.datadoghq.com/tracing/trace_pipeline/ingestion_mechanisms/?tab=java#in-the-agent .

Author: Added a link

You can also configure the sampling rate by service: see keep more traces for specific services

#### How to configure head-based sampling ?

Sampling rates are calculated to target 10 complete traces per second per Agent. You can increase or decrease this target by configuring the Datadog Agent parameter `max_traces_per_second` or the environment variable `DD_APM_MAX_TPS`.
Read more about this [ingestion mechanisms][5] in the documentation.

> **Note:** Changing the Agent configuration will impact the percentage sampling rates applied for all services reporting traces to the Datadog Agent.

Reviewer 5: ```suggestion
**Note:** Changing an Agent configuration will impact the percentage sampling rates applied for all services reporting traces to this Datadog Agent.
```

> For most scenarios, relying on the Agent-level configuration is enough to stay within the allotted quota and have enough visibility on what’s happening on your applications to make appropriate decisions for your business. 

Reviewer 5: ```suggestion
For most scenarios, relying on the Agent-level configuration is suited to stay within the allotted quota and have enough visibility on your applications performance and make appropriate decisions for your business. 
```

You can configure sampling rules by setting the environment variable `DD_TRACE_SAMPLING_RULES`. 

For example, to send `20%` of the traces for the service named `my-service`:

```
DD_TRACE_SAMPLING_RULES=[{"service": "my-service", "sample_rate": 0.2}]
```

Read more about this [ingestion mechanisms][6] in the documentation.

### Keep more errors traces

#### Why ?

Traces with error spans are often symptoms of system failures. Keeping a higher proportion of transactions with errors ensures to always be able to investigate example individual requests.

#### What solutions does Datadog provide ?

In addition to head-based sampled traces, each Agent keeps additional error spans out of the traffic **not captured** by head-based sampling. 

{{< img src="/tracing/guide/ingestion_sampling_use_cases/error_sampling.png" alt="Error Sampling" style="width:100%;" >}}

**Note:** Distributed pieces of the trace chunks might not be ingested as the sampling happens locally at the Datadog Agent level.

#### How to configure error sampling ?

You can configure the number of error chunks per second per Agent that you want to capture by setting the environment variable `DD_APM_ERROR_TPS`.

> To ensure to ingest **all errors**, set the Agent parameter to an arbitrary high value. To disable the error sampler, set `DD_APM_ERROR_TPS` to `0`.

Reviewer 5: Do you want to tell the default value also?

---
title: Trace Sampling Use Cases
kind: guide
further_reading:
- link: "/tracing/guide/trace_ingestion_volume_control/"
  tag: "Guide"
  text: "How to control ingested volumes"
---

## Overview

> Trace data tends to be repetitive. A problem in your application is rarely identified in a single trace. For high throughput services, particularly for incidents that require your attention, an issue shows symptoms in multiple traces. Consequently, there’s usually no need for you to collect every single trace for a service or endpoint. Datadog APM [ingestion control mechanisms][1] help you keep the visibility that you need to troubleshoot problems while cutting down the noise and managing costs.

Reviewer 4: ```suggestion
Trace data tends to be repetitive, and a single trace is rarely enough to fully debug an issue. For high throughput services, particularly for incidents that require your attention, crucial data is spread across multiple traces. Consequently, there’s usually no need for you to collect every single trace for a service or endpoint, or every span within a trace. Datadog APM [ingestion control mechanisms][1] provides the visibility that you need to troubleshoot problems while cutting down the noise and managing costs.
```

Reviewer 2: Not sure I agree with this rewrite. The information is not "spread across multiple traces"; it's that a significant problem will show up the same way, repeatedly" in multiple traces, so it doesn't matter which instance of the problem traces you ingest.

Reviewer 2: modified a bit

Reviewer 4: Ah I understand what this section is saying now. I found it confusing at first, and thought it was alluding to multiple traces containing the data.

> Ingestion mechanisms are controllable from the Datadog Agent and from Datadog tracing libraries. If you are using OpenTelemetry SDKs to instrument your applications, read [Ingestion Sampling with OpenTelemetry][2].

Reviewer 4: ```suggestion
Ingestion mechanisms are configuration within the Datadog Agent and from Datadog tracing libraries. If you are using OpenTelemetry SDKs to instrument your applications, read [Ingestion Sampling with OpenTelemetry][2].
```

This guide helps you understand when and how to use ingestion control configurations depending on the main use cases you might encounter. It covers:

- [Determining which ingestion mechanisms are used](#determining-which-ingestion-mechanisms-are-used) for a given service
- [Use cases that focus on keeping particular types of traces](#keeping-certain-types-of-traces)
- [Use cases that focus on reducing ingested traces](#reducing-ingestion-for-high-volume-services)

## Determining which ingestion mechanisms are used

> To identify what ingestion mechanisms are currently used in your Datadog environment, navigate to the [Ingestion Control Page][3].

Reviewer 4: ```suggestion
To identify which ingestion mechanisms are currently used in your Datadog environment, navigate to the [Ingestion Control Page][3].
```

{{< img src="/tracing/guide/ingestion_sampling_use_cases/ingestion_control_page.png" alt="Ingestion Control Page" style="width:90%;" >}}

The table gives insights on ingested volumes *by service*. The Configuration column provides a first indication of the current set up. It shows: 

> - `AUTOMATIC` if the sampling rate calculated in the Datadog Agent is applied to the traces that start from the service. Read more about the specifics of [Datadog Agent ingestion logic][5].

Reviewer 4: I'm not sure what "from the service" is referring to here and in the bullet below.

Reviewer 2: I think it means that the call to the service is the root span of the trace... the trace starts from that service.

- `CONFIGURED` if a custom sampling rate configured in the tracing library is applied to the traces that start from the service.

Click on services to see details about what sampling decision makers are used for each service, as well as what [ingestion sampling mechanisms][1] are leveraged for ingested spans' services. 

{{< img src="/tracing/guide/ingestion_sampling_use_cases/service_ingestion_summary.png" alt="Service Ingestion Summary" style="width:90%;" >}}

## Keeping certain types of traces

### Keeping entire transaction traces

#### Why ? 

> Ingesting entire transaction traces ensures to keep visibility over the **end-to-end service request flow** for specific individual requests.

Reviewer 4: ```suggestion
Ingesting entire transaction traces ensures visibility over the **end-to-end service request flow** for specific individual requests.
```

#### What solutions does Datadog provide ?

> Complete traces can be ingested with [head-based sampling][4] mechanisms: the decision of keeping or dropping the trace is taken at the trace creation (trace root span), and this decision is propagated in the request context to downstream callee services.

Reviewer 4: ```suggestion
Complete traces can be ingested with [head-based sampling][4] mechanisms: the decision to keep or drop a trace is determined from the first span of the trace, or the "head", when the trace is created. This decision is propagated through the request context to downstream services.
```

{{< img src="/tracing/guide/ingestion_sampling_use_cases/head_based_sampling_keep.png" alt="Head-based Sampling" style="width:100%;" >}}

To decide which traces to keep and drop, the Datadog Agent automatically computes sampling rates for each service to apply at the trace creation, based on the application traffic.
- For low-traffic applications, a sampling rate of 100% is applied

> - For high-traffic applications, a lower sampling rate is applied

Reviewer 4: "a lower sampling rate is applied"
Can we be more specific here? Are there defaults?

You can also override Agent sampling rate by configuring the sampling rate by service. See how to [keep more traces for specific services](#keep-more-traces-for-some-specific-services--resources-for-better-visibility) for more information.

> #### How to configure head-based sampling ?

Reviewer 4: ```suggestion
#### How to configure head-based sampling
```

> Sampling rates are calculated to target 10 complete traces per second per Agent. You can increase or decrease this target by configuring the Datadog Agent parameter `max_traces_per_second` or the environment variable `DD_APM_MAX_TPS`.

Reviewer 4: ```suggestion
Sampling rates are calculated to target 10 complete traces per second per Agent by default. You can increase or decrease this target by configuring the Datadog Agent parameter `max_traces_per_second` or the environment variable `DD_APM_MAX_TPS`. Note that this is a **target** number of traces and is the result of averaging traces over a period of time. It is *not* a hard limit, and traffic spikes could cause significantly more traces to be sent to Datadog for short periods of time.
```

> Read more about this [ingestion mechanisms][5] in the documentation.

Reviewer 4: ```suggestion
Read more about this [ingestion mechanism][5] in the documentation.
```

**Note:** Changing an Agent configuration will impact the percentage sampling rates applied for all services reporting traces to this Datadog Agent.

For most scenarios, relying on the Agent-level configuration is suited to stay within the allotted quota and have enough visibility on your applications performance and make appropriate decisions for your business. 

### Keeping more traces for high visibility into specific services or resources

#### Why ?

If some services and requests are critical to your business, you may want to send all related traces to Datadog to be able to look into any of the individual transactions.

#### What solutions does Datadog provide ?

> By default, sampling rates are calculated to target 10 traces per second per Datadog Agent. You can override the default calculated sampling rate by configuring **[sampling rules][6]** in the tracing library. 

Reviewer 4: Doesn't this repeat line 59?

You can configure sampling rules by service. For traces starting from the service targeted by the sampling rule, the defined percentage sampling rate will be applied instead of the default sampling rate calculated by the Datadog Agent.

#### How to configure a sampling rule ?

You can configure sampling rules by setting the environment variable `DD_TRACE_SAMPLING_RULES`. 

For example, to send `20%` of the traces for the service named `my-service`:

```
DD_TRACE_SAMPLING_RULES=[{"service": "my-service", "sample_rate": 0.2}]
```

> Read more about this [ingestion mechanisms][6] in the documentation.

Reviewer 4: ```suggestion
Read more about this [ingestion mechanism][6] in the documentation.
```

This guide helps you understand when and how to use ingestion control configurations depending on the main use cases you might encounter. It covers:

- [Determining which ingestion mechanisms are used](#determining-which-ingestion-mechanisms-are-used) for a given service
- [Use cases that focus on keeping particular types of traces](#keeping-certain-types-of-traces)
- [Use cases that focus on reducing ingested traces](#reducing-ingestion-for-high-volume-services)

## Determining which ingestion mechanisms are used

To identify which ingestion mechanisms are currently used in your Datadog environment, navigate to the [Ingestion Control Page][3].

{{< img src="/tracing/guide/ingestion_sampling_use_cases/ingestion_control_page.png" alt="Ingestion Control Page" style="width:90%;" >}}

The table gives insights on ingested volumes *by service*. The Configuration column provides a first indication of the current set up. It shows: 
- `AUTOMATIC` if the sampling rate calculated in the Datadog Agent is applied to the traces that start from the service. Read more about the specifics of [Datadog Agent ingestion logic][5].

> - `CONFIGURED` if a custom sampling rate configured in the tracing library is applied to the traces that start from the service.

Author: ```suggestion
- `CONFIGURED` if a custom trace sampling rate configured in the tracing library is applied to the traces that start from the service.
```

Click on services to see details about what sampling decision makers (for example Agent or tracing library, rules or sample rates) are used for each service, as well as what [ingestion sampling mechanisms][1] are leveraged for ingested spans' services. 

{{< img src="/tracing/guide/ingestion_sampling_use_cases/service-ingestion-summary.png" alt="Service Ingestion Summary" style="width:90%;" >}}

In the Service Ingestion Summary example above, the **Ingestion reasons breakdown** table shows that most of the ingestion reasons for this service come from `rule` ([user defined sampling rule][6]).

> The Top sampling decision makers for this service show that web-store service gets sampling decisions from `web-store`, `shopist-web-ui`, `shipping-worker`, `synthetics-browser`, and `product-recommendation`. These five services all contribute in the overall sampling decisions that affect the `web-store` service spans. When determining how to fine tune the ingestion for web-store, all five services need to be considered.

Reviewer 2: ```suggestion
The Top sampling decision makers for this service show that `web-store` service gets sampling decisions from `web-store`, `shopist-web-ui`, `shipping-worker`, `synthetics-browser`, and `product-recommendation`. These five services all contribute in the overall sampling decisions that affect the `web-store` service spans. When determining how to fine tune the ingestion for web-store, all five services need to be considered.
```
